### PR TITLE
Feature/us 28/product filtering feature

### PR DESCRIPTION
**Pull request by victor-villca**
#### What did I do?
I have implemented the Product Filtering Feature as per the requirements outlined in user story ID 28. This includes the following functionalities:

1. **Filter by Price:** Users can now filter products based on price ranges.
2. **Filter by Category:** The system allows users to filter products by category.
3. **Filter by Type:** Users can filter products based on types.
4. **Filter by Size:** The system supports filtering products by size.
5. **Filter by Color:** Users can now filter products by color.

Additionally, I have implemented the ability for users to apply multiple filters simultaneously, allowing for a more refined shopping experience. For example, users can filter products that are specifically for men within a certain price range.

The Filter UI is designed to be responsive, ensuring a seamless experience for users on mobile devices.

#### How did I do it?
I implemented these features by enhancing the existing product search and retrieval logic. The backend now supports querying based on various parameters, and the frontend has been updated to provide a user-friendly interface for applying and managing filters.

To achieve responsiveness, I utilized responsive design principles, ensuring that the Filter UI adapts appropriately to different screen sizes and orientations.

#### Why did I do it?
The goal is to provide users with a tailored shopping experience, allowing them to easily find products that match their preferences. The advanced filtering system enhances the overall usability of the platform and caters to users who have specific criteria in mind when searching for products.

**Author:**  @victor-villca 
**Code Reviewer:**  @Karvlox 
**Tester:** @denisgandel 
**Clickup: https://app.clickup.com/t/8686a783g**

### CODE REVIEW

**Comment by denisgandel** [comment]
https://github.com/Les-Jailes/Boutique_Clothing/assets/135136502/1cfc2524-7e96-4e9b-b31b-f88548ac05a2

When I select a genre, it shows me all the products that are related to that genre, but when I select a category, it shows me not only for the chosen genre, but for all the genres there are, even products that are not of that type.

**Comment by victor-villca** [comment]
> Video.sin.titulo.Hecho.con.Clipchamp.mp4 
> When I select a genre, it shows me all the products that are related to that genre, but when I select a category, it shows me not only for the chosen genre, but for all the genres there are, even products that are not of that type.

That is not a bug, it is feature.

**Comment by denisgandel** [comment]
https://github.com/Les-Jailes/Boutique_Clothing/assets/135136502/8274d0f5-3a80-40b7-9f5f-a1c1215a936d

Your filters do not work correctly, because they should return only the products that comply with the selected filters, since it still shows products that do not belong to the type of the selected filter, or the selected color or the selected genre, and some products start to duplicate.

**Comment by denisgandel** [comment]
> > Video.sin.titulo.Hecho.con.Clipchamp.mp4
> > When I select a genre, it shows me all the products that are related to that genre, but when I select a category, it shows me not only for the chosen genre, but for all the genres there are, even products that are not of that type.
> 
> That is not a bug, it is feature.

If it would be a bug, because the functionality of the filter should be that it simply shows all the elements that do comply with the selected filters or aspects, and not show you all the elements that already minimally comply with a filter, besides that is the expected functionality for a filter.

**Comment by denisgandel** [comment]
https://github.com/Les-Jailes/Boutique_Clothing/assets/135136502/2e2972b2-f6dd-4b55-8404-4c75fd66a7b1

In addition, when the products are duplicated, they break the page layout, because the page layout should have a maximum of 12 products per page, but due to the duplication of products in this example we can see the following order:

* 1 page: 12 products
* 2 page: 18 products
* 3 page: 23 products
* 4 page: 22 products

**Comment by denisgandel** [comment]
Also, it would be nice that when the customer selects a gender, then the size section is updated to show only the sizes depending on the gender.

File: src/app/pages/products/page.jsx [hunk 1]
```diff
@@ -5,28 +5,163 @@ import createPagination from "@/utils/Pagination";
...
+      colorOrSizeFilter =
+        filterByColorOrSize(product, checkedLabels, "color") &&
+        filterByColorOrSize(product, checkedLabels, "size");
+    }
+
+    if (checkedLabels.price && checkedLabels.price.length > 0) {
+      priceFilter = filterByPrice(product, checkedLabels);
+    }
+
+    // console.log(
```
Comment (Karvlox) [inline src/app/pages/products/page.jsx]: Remove the commented code

**Review by Karvlox — APPROVED** [review]
Good job

**Comment by victor-villca** [comment]
> > > Video.sin.titulo.Hecho.con.Clipchamp.mp4
> > > When I select a genre, it shows me all the products that are related to that genre, but when I select a category, it shows me not only for the chosen genre, but for all the genres there are, even products that are not of that type.
> > 
> > 
> > That is not a bug, it is feature.
> 
> If it would be a bug, because the functionality of the filter should be that it simply shows all the elements that do comply with the selected filters or aspects, and not show you all the elements that already minimally comply with a filter, besides that is the expected functionality for a filter.

This bug was solved with the latest commit

File: src/app/pages/products/page.jsx (hunk 1 above)
Comment (victor-villca) [reply src/app/pages/products/page.jsx]: Ok

**Review by saletG08 — APPROVED** [review]
I don't have comments, good job

**Pull request merged**